README.md
# 3D-Printer-Enclosure
### Smooth PID controlled heated 3d printer enclosure control w/ display & Octoprint interface.

## Features:
* Simple 3 button navigation (up/down/select) w/ long press sub-menus
* Displays: enclosure temp, room temp, cooling fan %, heater status, cooldown status
* 3 fan channels for fresh air cooling, internal air mixing, and heater
* PID/PWM Fresh air cooling fan control
* PWM Mixing Fan Control w/ Manual Override
* Fan kickstart
* Anti fan stall (obeys minimum fan speeds)
* Heater Control (with hysteresis + min switching time)
* Heater temp offset (prevents heating/cooling "battles")
* Heat Soak Monitor (Auto Heater Off)
* Heater timer (for cold climates/weaker heaters, when heat soak does not occur)
* Case LED PWM Control
* Spotlight LED PWM Control
* Cooldown Mode
* Dallas swap #define (no messing with 1-wire addresses)
* Several other #defines to make customizing & tuning easier
* *NEW* Octoprint heat/auto/cool mode triggering

## Hardware:
 - Arduino Pro-mini 328p 16MHz/5V (or similar)
 - 2x Dallas 1-wire temp sensors (enclosure and room temps)
 - *NEW* 1602 i2c LCD (default)
 or
 - 20x2 IEE serial VFD display (softserial)
 - 3x logic level MOSFETs (for the 2 fan outputs and 1 LED output)
 - 2 or more 12V fans (at least 1 for fresh air, at least 1 for mixing, and 1 for the heater if used)
 - LEDs for case lighting
 - High powered LED spotlight (Luxeon Star or similar)
 - Power LED driver (Meanwell LDD-350LL or similar 5V PWM dimmable device)
 - 3x SPST momentary NO tactile buttons (6mm square type if using my printed VFD enclosure)
 - 5VDC/110VAC 40A SSR (to switch the heater)
 - ~250W personal space heater (120VAC ceramic type preferred)
 - 50A SSR (for the heater)

## Operation-----------------------------------------------------------------------

### Auto & AutoH mode:
Boots to auto mode, with heater function disabled. In auto mode, the fresh fan controlled by PID to maintain tempDesired. If heating is enabled, it is hysteresis controlled to heat the enclosure up to tempDesired minus heaterTempOffset, plus or minus heaterTempHysteresis. The heater function automatically turns off if heaterMaxOffTime is elapsed since the heater was turned off. This is an easy way to determine if the enclosure is heat soaked, then heating is automatically disabled to prevent continued heating after a print is finished. If the heater timer feature is used, the heater will turn off after the set hours has elapsed. The LCD displays "Auto" when heater feature is disabled, or "AutoH" when heater is enabled, the actual/set temperatures, room temp, and the fan output % (or "Heating").Short pressing the Up/down buttons will adjust tempDesired, and the select button will switch to cooldown mode.

### Cooldown mode:
In cooldown mode, the heater remains off and fresh fan spins 100% until the temperature falls below tempCooldown. Hysteresis is used to prevent frequent fan cycling in cooldown mode. Also, in this mode the display shows "Cool", actual/set temperatures, Room temp, and "Cooling" or "Idle" depending on fan state. Short pressing the up/down buttons will adjust tempCooldown, and short pressing  the select button will switch to auto mode.

### Mixing fan control:
The mixing fan runs at a manually set speed if that speed is greater than the fresh air fan speed or minimum mix fan speed. If the fresh air fan speed exceeds the manual speed setting, the manual mixing fan speed is overridden, and the mixing fan will run at the fresh air fan speed. The mixing fan turns 100% on whenever the heater is on, and stays on for heaterMixFanDelay millis after the heater turns off.

### Fan kickstart & anti-stall:
All fans have a kickstart feature, which initially sends extra power to the fans to get them moving. Also no fans will startup (or continue to buzz) unless their input pwm is enough to keep them spinning. This combination of kickstart and minimum PWM control reduces PID reduced delay and overshoot, and prevents unnecessary buzzing/heating of non-moving fans. StartupPWM and MinPWM compiler defines can be used to tune behavior of both fresh and mixing fans. Adjust them as needed to match your choice of fans.

### Changing manual mix fan speed:
Long pressing (>2sec default) the up button switches to the manual mixing fan speed setting. Press select to return to auto or autoh mode.

### Changing LED brightness:
Long pressing the down button switches to the case led setting. In case led brightness mode, pressing up or down will adjust the case LED brightness.
Press select to advance to the spotlight setting menu, where up/down adjusts spotlight brightness. Press select again to return to auto or autoh mode.

### Changing Heater settings (if a heater is used):
Long pressing select goes to the heater enable/disable setting. Pressing up enables heating, and down disables it. This setting will persist until the next reboot, if the heat soak timer is triggered, or if the heater timer has elapsed. Short pressing select then goes to the heater timer setting. If left at 0hrs, the heater timer function is disabled (thus the heater function will remain on until heat soak is sensed, it is manually turned off, or a reboot). Pressing up/down increases/decreases the timer by 1 hour, from 0 to heaterTimerMax. Any time the heater hours is adjusted, the heat timer is restarted at t=0. Short pressing select again returns to auto or autoh mode.

*While working in any long press menu, the fans and heater will be controlled as if in auto or autoh mode.*

*The desired & cooldown temperature adjustments are limited to max/min values for safety. Change them as needed to better suit your setup.*

### Octoprint integration
Optional octoprint integration allows auto, heat, and cool modes to be controlled by 2 pins connected to the RPi. The OP inputs are read periodically in all modes. If NOT in a UI input mode while the inputs change, the code switches to the appropriate mode (with default heater hours for heat mode).
Manual input is supreme... OP changes are ignored if they occur during GUI input modes.

## Pinouts
Arduino pin | Connect to
----------- | ----------
D2  |  Dallas temp data pins
D3  |  Fresh air fan MOSTFET input
D4  |  Select button (to gnd)
D5  |  Up button (to gnd)
D6  |  Down button (to gnd)
D9  |  Mixing fan MOSTFET input
D10  |  Case lights MOSFET input
D11  |  Spotlight LED driver input
D13  |  Heater SSR input
A4  |  LCD SDA
A5  |  LCD SCL
A0  |  Octoprint heat signal
A1  |  Octoprint cool signal
*D8*  |  NC (reserved if using a VFD)
*D12*  |  VFD Rx pin (pin14 on VFD)

*VFD display uses D8 & D12 instead of A4 & A5*

## Installation
This code is written for compatibility with PlatformIO. Edit configuration.h then upload to your uC.


changelog.md
3D Printer Enclosure Change-log

# 0.1.1 Current Development
Add feature: Octoprint inputs
Added hardware: 1602 i2c LCD (untested)

# 0.1 Initial Release
3 button 20x2 vfd interface
PID mixing and cooling fans
Heat, auto, and cooldown modes
Case & Spotlight controls


src/ieeVfdCommands.h
// ieeVfdCommands.h
// by: truglodite
// updated: 6/5/2019
//
////////////////////////////////////////////////////////////////////////////////
// Control bytes for the IEE VFD in "LCD" or "Intel" mode (IEE S036x2_N manual,
// hex converted to decimal). Most of these codes have been tested on an
// IEE 036X2-122-09220 VFD (AKA 05464ASSY35119-02A). Other IEE vfd's are
// likely to work.
////////////////////////////////////////////////////////////////////////////////
const uint8_t VFDCTRL_CURSOR_BACKSPACE = 8;
const uint8_t VFDCTRL_CURSOR_ADVANCE = 9;
const uint8_t VFDCTRL_LINE_FEED = 10;
//const uint8_t VFDCTRL_CURSOR_BLINK_BLOCK = 11; // Applies to 036X2–151–05240 & 036X2-160-05440 only.
//const uint8_t VFDCTRL_CURSOR_UNDERBAR = 12; // Applies to 036X2–151–05240 & 036X2-160-05440 only.
const uint8_t VFDCTRL_CARRIAGE_RETURN = 13;
const uint8_t VFDCTRL_CURSOR_OFF = 14;
const uint8_t VFDCTRL_CURSOR_ON = 15; //default
const uint8_t VFDCTRL_SCROLL_LINE_LOCK = 16; // 2 byte command...
/* 0   Locks line 0
 * 1   Locks lines 0 and 1
 * 2   Locks lines 0, 1 and 2
 * 255   Cancel Line Lock
 */
const uint8_t VFDCTRL_SET_VERTICAL_SCROLL_MODE = 17; //default auto cr+lf
const uint8_t VFDCTRL_SET_HORIZONTAL_SCROLL_MODE = 19;
const uint8_t VFDCTRL_SOFTWARE_RESET = 20;
const uint8_t VFDCTRL_CLEAR_DISPLAY_HOME_CURSOR = 21;
const uint8_t VFDCTRL_HOME_CURSOR = 22;
const uint8_t VFDCTRL_SET_DATABIT_7_HIGH = 23; //for the following byte
const uint8_t VFDCTRL_BEGIN_USER_DEFINED_CHARACTER = 24; //7 byte command follow with...
/* 1 User defined character storage location byte (10 available, from 246 - 255)
 *  and
 * 5 bytes of data for the character dots. 0=Dot off, 1=Dot on.
 * ex: "24 246 P P P P P"
 * where P are position bytes containing 8 bits each (01101010)
 *
CHARACTER DOT MATRIX (physical location of the actual dots)
1   2   3   4   5
6   7   8   9   10
11  12  13  14  15
16  17  18  19  20
21  22  23  24  25
26  27  28  29  30
31  32  33  34  35

Map of data bits for part#'s (036X2-100,-105,-122,-124,-134,-130)
Bit>  7   6   5   4   3   2   1   0
Btye  ------------------------------
3     33  15  34  16  35  17  0   18
4     29  11  30  12  31  13  32  14
5     25  07  26  08  27  09  28  10
6     21  03  22  04  23  05  24  06
7     0   0   0   0   19  01  20  02

Map of data bits for part#'s (036X2-106,-120,-121,-151,-116,-160)
Bit>  7   6   5   4   3   2   1   0
Byte  ------------------------------
3     29  20  11  02  28  19  10  01
4     31  22  13  04  30  21  12  03
5     33  24  15  06  32  23  14  05
6     35  26  17  08  34  25  16  07
7     0   0   0   0   0   27  18  09
 */
const uint8_t VFDCTRL_SET_ADDRESSBIT_0_HIGH = 25; //must be followed by control codes below
const uint8_t VFDCTRL_CURSOR_UP_ONE_LINE = 26;
const uint8_t VFDCTRL_CURSOR_MOVE_TO_LOCATION = 27; //2 byte command...
//Follow with a 1 byte position ID. Screen positions are numbered from left to right, top to
//bottom starting with 0.
const uint8_t VFDCTRL_CHARSET_EUROPEAN = 28; //default
const uint8_t VFDCTRL_CHARSET_KATAKANA = 29;
const uint8_t VFDCTRL_CHARSET_CRYLLIC = 30;
const uint8_t VFDCTRL_CHARSET_HEBREW = 31;

//***THE FOLLOWING VFD CONTROL CODES MUST BE PRECEDED BY the byte '25' (Set Ao high)***
const uint8_t VFDCTRL_SET_SCREEN_OR_COLUMN_BRIGHTNESS = 48; //3 byte command...
/* Follow with Column ID and Brightness Level.
 * Column ID 0-X???, use 255 for entire screen
 * Brightness Level brightest=0 to dimmest=7. default=0
 */
const uint8_t VFDCTRL_BEGIN_BLINKING_CHARACTERS = 49; //2 byte command...
/* Follow with rate and type byte
 * Type       Rate    byte
 * -------------------------------------
 * Character  OFF     00   <----default
 * Character  1hz     01
 * Character  2hz     02
 * Character  4hz     04
 * Underbar   OFF     96  -applies only to 36X2–151–05240
 * Underbar   1hz     97  -" "
 * Underbar   2hz     98  -" "
 * Underbar   4hz     100 -" "
 * Both       OFF     128 -" "
 * Both       1hz     129 -" "
 * Both       2hz     130 -" "
 * Both       4hz     132  -" "
 */
const uint8_t VFDCTRL_END_BLINKING_CHARACTERS = 50;
const uint8_t VFDCTRL_BLANK_DISPLAY_ON = 51;
const uint8_t VFDCTRL_BLANK_DISPLAY_OFF = 52;
//static const char VFDCTRL_COMMA_PERIOD_TRIANGLE_FUNCTION = 53; //Applies to 036X2-121-11120 only.
const uint8_t VFDCTRL_ERASE_LINE_WITH_END_BLINK = 54; //2 byte command...
/*
 *
 */
const uint8_t VFDCTRL_SET_CR_LF_DEFINITIONS = 55; //2 byte command...
/* Follow with the definition byte
 * bits LFin   CRin
 * --------------------------------
 * 0    LF     CR     <---default
 * 1    LF+CR  CR
 * 2    LF     CR+LF
 * 3    LF+CR  CR+LF
 */
//static const char VFDCTRL_UNDERBAR_ON = 56; // Applies to 036X2–151–05240 & 036X2-160-05440 only.
//static const char VFDCTRL_UNDERBAR_OFF = 57; // Applies to 036X2–151–05240 & 036X2-160-05440 only.
const uint8_t VFDCTRL_SELECT_RIGHT_TO_LEFT_DATA_ENTRY = 58;
const uint8_t VFDCTRL_SELECT_LEFT_TO_RIGHT_DATA_ENTRY = 59;  //default
const uint8_t VFDCTRL_SCREEN_SAVER_ON = 60;
const uint8_t VFDCTRL_SCREEN_SAVER_OFF = 61; //default
const uint8_t VFDCTRL_SELF_TEST_EXECUTE = 62;
const uint8_t VFDCTRL_SELF_TEST_TERMINATE = 63;


src/configuration.h
// configuration.h
// by: Truglodite
// updated: 6-6-2019
//
// User configuration for printer_enclosure.ino
//
///////////////////////////////////////////////////////////////////////////////

// Uncomment if your room and enclosure temperatures are swapped.
//#define swapDallas


#define octoprintControl  // Uncomment if using octoprint triggering
  // Note, octoprint input changes will trigger a mode change ONLY IF they
  // occur during an automated mode (auto, cooldown, and heating). IE, if a
  // user is changing a setting in a GUI input mode, the OP command will be
  // ignored, and the user's end settings will carry on until one of the pins
  // changes again.
#ifdef octoprintControl
  #define octoprintHeatPin  A0  // triggers 'heater on' when high
  #define octoprintCoolPin  A1  // triggers 'cooling on' when high
  // triggers 'auto on' when both heat and cool pins are low
  #define octoprintUpdateRate  3000  // millis between reading op input pins
#endif


#define vfd_display  // Uncomment if using a 2x20 IEE VFD display
#ifdef vfd_display
  #include "ieeVfdCommands.h"
  #include <SoftwareSerial.h>
  #define VFD_RX       8    // not used, but required for softserial
  #define VFD_TX       12   // connect to VFD serial rx (pin 14)
#endif

// A 1602 i2c LCD display is default
// defaults... A4 = SDA, A5 = SCL
#ifndef vfd_display
  #include <LiquidCrystal_I2C.h>
  #include <Wire.h>
#endif

// Pinouts /////////////////////////////////////////////////////////////////////
#define dallasPin    2    // To Dallas data pins (requires one 4k7 pullup)
#define freshFanPin  3    // to cooling fan mosfet
#define selectPin    4    // buttons are momentary N.O. (internal pullups)
#define upPin        5
#define downPin      6
#define mixFanPin    9    // to mixing fan mosfet
#define caseLedPin   10   // to case light mosfet
#define spotlightPin 11   // to spotlight LED driver (LDD-350LL)
#define heaterPin    13   // to heater ssr input

// General preferences ///////////////////////////////////////////////
#define debounce     50   //button debounce millis (no repeaters, so this can be short)
#define longPress    2000 //millis to hold button for special menus
#define splashTime   3000 //millis to show splash screen
#define displayPeriod 200 //millis between display updates (<200 values may blur)

#define caseLedPWMdefault 255 //bootup case lighting brightness
#define caseLedIncrement 25.5 //PWM change per up/down press
#define spotlightPWMdefault 255 //bootup spotlight brightness
#define spotlightIncrement 25.5 //PWM change per up/down press

#define freshFanStartupPWM 50 //fresh fan PWM value used for kickstarting
#define freshFanMinPWM   13//fresh fan PWM below which the fan stops spinning

#define mixFanPWMdefault 0 //bootup manual mix fan speed
#define mixFanIncrement 25.5 //PWM change per up/down press
#define mixFanStartupPWM 50  //mix fan PWM value used for kickstarting
#define mixFanMinPWM   15  //mix fan PWM below which the fan stops spinning

#define fanStartupPeriod 500 //millis to kickstart all fans

#define heaterMinSwitchTime 10000  //min millis between heater switching
#define heaterTempOffset 1.0  //celsius below desired temp when heater is allowed on (to avoid heater vs cooling fan fights)
#define heaterHysteresis 0.5 //Celsius hysteresis to use for heater
#define heaterMixFanDelay 30000 //millis to keep mixing fan on after heater turns off (>heaterMinSwitchTime)
#define heaterMixFanPWM 255 //pwm value for mix fan while heater is on (255)
#define heaterMaxOffTime 35000 //millis of heater off time to allow before shutting down heating function (heatsoak timer... >heaterMixFanDelay)
#define heaterHoursIncrement 1 //hrs to increment heater timer
#define heaterHoursMax 48 //max hrs allowed for heater timer (any int... 48hrs should cover most prints)
#define heaterHoursDefault 2

#define tempDefault  44.0   //bootup desired Celsius setting (50)
#define tempMinSetting      20.0   //min desired t allowed(20)
#define tempMaxSetting      50.0   //max desired t allowed (50). Steppers may fail at >50C.
#define tempCooldownDefault 24.0 //default cooldown temperature (30)
#define tempCooldownHysteresis 1.0 //hysteresis for fans in cooldown mode
#define tempIncrement 2.0   //celsius change per up/down push (2)

#define dallasResolution 12 //dallas bit resolution (4,8, or 12)
#define sizeOfBuffer 3    //the number of t readings to store for averaging (1 = disable, larger buffers use more memory
                          //and may result in laggy PID feedback (increases overshoot)
#define tempPeriod   800  //millis to wait for a requested temp (>760 for 12bit)

// PID tuning parameters... will depend on many aspects of your setup.
// Use ZN or other manual tuning methods. The default PID's work well
// with the code author's enclosure using a specific assortment of hardware,
// but they should be a good starting point for tuning most setups.
#define freshFanPIDsampleTime 200  //200ms should work for most... beware that PID values below should scale with the sample time
double freshFanP=    10;  //the pid's must be set whether or not you use bang bang
double freshFanI=    1;   //default p,i,d = 10,5,1
double freshFanD=    1;
